File: 138.Copy-List-with-Random-Pointer/list.py
from typing import Optional


class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: Optional[Node]) -> Optional[Node]:
        table = {}
        it = head

        # O(n)
        while it:
            table[it] = Node(it.val)
            it = it.next

        # O(n)
        it = head
        while it:
            node = table[it]

            node.next = table.get(it.next)
            node.random = table.get(it.random)

            it = it.next

        return table.get(head)

File: 206.Reverse-Linked-List/list.rs
impl Solution {
    pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut curr = head;
        let mut prev = None;

        // O(n)
        while let Some(mut node) = curr {
            curr = node.next;
            node.next = prev;
            prev = Some(node);
        }

        prev
    }
}
